<launch>
  
  <arg name="video_device" default="/dev/video0" />
  <arg name="image_width" default="640" />
  <arg name="image_height" default="480" />
  <arg name="pixel_format" default="mjpeg" />
  <arg name="frame_rate" default="30" />
  <arg name="camera_name" default="camera" />
  <!-- <arg name="camera_frame_id" default="camera" /> -->
  
  <!-- Use different namespaces for the 2nd camera or greater -->
  <group ns="$(arg camera_name)">
    
    <!-- USB camera driver + controller manager -->
    <node name="hardware" pkg="nodelet" type="nodelet" 
          args="standalone usb_cam_hardware/USBCamHardware" output="screen">
      <remap from="camera_info_controller/camera_info" to="camera_info" />
      <remap from="mjpeg_controller/image" to="image" />
      <remap from="compressed_packet_controller/packet" to="packet/compressed" />
      <param name="video_device" value="$(arg video_device)" />
      <param name="image_width" value="$(arg image_width)" />
      <param name="image_height" value="$(arg image_height)" />
      <param name="pixel_format" value="$(arg pixel_format)" />
      <param name="framerate" value="$(arg frame_rate)" />
    </node>
    
    <!-- Controller starter that asks the manager to start -->
    <node name="controller_starter" pkg="controller_manager" type="controller_manager" 
          args="spawn camera_info_controller compressed_packet_controller mjpeg_controller" output="screen" />
    <!-- Recommendation: use controllers you actually required -->
    <!--   camera_info_controller publishes camera infos synchronized with images -->
    <!--   compressed_packet_controller publishes packets from the camera (so requires subscriber-side decoding) -->
    <!--   mjpeg_controller does server-side decoding (usually heavy) and publishes images -->
    
    <!-- Parameters for controllers -->
    <param name="camera_info_controller/type" value="usb_cam_controllers/CameraInfoController" />
    <param name="camera_info_controller/camera_frame_id" value="$(arg camera_name)" />
    
    <param name="compressed_packet_controller/type" value="usb_cam_controllers/CompressedPacketController" />
    <param name="compressed_packet_controller/format" value="jpeg" />
    <!-- pub 1 packet and then skip 2 packets -->
    <!-- <param name="compressed_packet_controller/skip" value="2" /> -->
    
    <param name="mjpeg_controller/type" value="usb_cam_controllers/MjpegController" />
    
  </group>
  
</launch>
